book/analysis/performance-of-python-types.md

```diff
@@ -5,90 +5,74 @@ position: 4
-we explore an alternative list implementation in a later section, we
-will see that this inefficiency is a concession that is made in order to
-allow $$O(1)$$ index operations.
-
-Much like popping from the start or middle of a list, inserting at an
-index is $$O(n)$$ in Python, as every subsequent element must be shifted
-one position closer to the end in order to accommodate the new element.
-Deletion unsurprisingly behaves in the same way.
-
-Iteration—somewhat intuitively—is $$O(n)$$. Iterating over $$n$$ items
-requires $$n$$ steps. This also gives us an intuitive understanding of
-why the `in` operator in Python is $$O(n)$$ when operating over a list—
-effectively we must iterate over every element of the list to determine
-if the element in question is contained therein.
-
-The slice operations require some more thought. To access a slice `[a:b]`
-of an array, we must iterate through every element between that at index
-`a` and that at index `b`. As such slice access is $$O(k)$$ where $$k$$
-is the number of elements in the slice. Deleting a slice is $$O(n)$$ for
-the same reason that deleting a single element is—$$n$$ subsequent
-elements must then be shifted toward the beginning of the list.
-
-As mentioned previously, concatenation is $$O(k)$$ where $$k$$ is the
-length of the concatenated list, due to an effective $$k$$ appends.
-It follows that multiplying a list is $$O(nk)$$ as multiplication of
-a $$k$$ length list $$n$$ times will involve $$k(n - 1)$$ appends, so
-in the order of $$nk$$.
-
-Reversing a list in $$O(n)$$ as each element must be repositioned.
-Finally, as previously mentioned [sort in Python](http://svn.python.org/view/python/trunk/Objects/listsort.txt?revision=69846&view=markup) is $$O(n log n)$$, the
-least intuitive of all of these outcomes, and [beyond the scope of this book](https://en.wikipedia.org/wiki/Timsort)
-to demonstrate.
-
-Below we provide a table of performance characteristics of Python list
-operations, for future reference.
+Python designers had many choices to make when they implemented the list data
+type. Each choice has a potential impact on how quickly the list can perform
+operations. One decision they made was to optimize the list implementation for
+commonly-used operations. This doesn't mean rarer operations are slow; it just
+means that common operations got preference during situations where the
+designers had to make a tradeoff.
+
+Two common operations are indexing and assigning to an index position. In
+Python lists, values are assigned to and retrieved from specific, known memory
+locations. No matter how large the list is, then, index lookup and assignment
+take a constant amount of time and are thus $$O(1)$$.
+
+Another common programming need is to grow a list. There are ways to do this:
+you can use the `append` method or the concatenation operator (`+`).
+
+The `append` method is $$O(1)$$ since the new value is stored at a specific
```

Comment:
I'm a little wary of being so handwavey with this. The "specific memory location" hasn't always been "already allocated". How about just a little more implementation detail here:

> The `append` method is "amortized" $$O(1)$$. In most cases, the memory required to append a new value has already been allocated, which is strictly $$O(1)$$. Once the C array underlying the list has been exhausted, it must be expanded in order to accommodate further `append`s. This periodic expansion process is linear in time to the size of the new array, which seems to contradict our claim that `append`ing is $$O(1)$$. However, the expansion rate is cleverly chosen to be 3x the previous size of the array, such that when we mentally spread the cost of expansion over each `append` that we are able to perform as a consequence, the cost per `append` is $$O(1)$$ _on an amortized basis_.